scipy/optimize/_simulated_annealing.py

```diff
@@ -0,0 +1,557 @@
+        # Local markov chain minimum energy and location
+        self.emin = state.current_energy
+        self.xmin = np.array(state.current_location)
+        # Global optimizer state
+        self.state = state
+        # Acceptance parameter
+        self.acceptance_param = acceptance_param
+        # Visiting distribution instance
+        self.visit_dist = visit_dist
+        # Wrapper to objective function and related local minimizer
+        self.obj_fun_wrapper = obj_fun_wrapper
+        self.not_improved_idx = 0
+        self.not_improved_max_idx = 1000
+        self._rand_state = rand_state
+        self.temperature_step = 0
+        self.K = 100 * len(state.current_location)
+
+    def run(self, step, temperature):
+        self.temperature_step = temperature / float(step + 1)
+        self.not_improved_idx += 1
+        for j in range(self.state.current_location.size * 2):
+            if j == 0:
+                self.state_improved = False
+            if step == 0 and j == 0:
+                self.state_improved = True
+            x_visit = self.visit_dist.visiting(
+                self.state.current_location, j, temperature)
+            # Calling the objective function
+            e = self.obj_fun_wrapper.func(x_visit,
+                *self.obj_fun_wrapper.fun_args)
+            self.obj_fun_wrapper.nb_fun_call += 1
+            if e < self.state.current_energy:
+                #  print('Better energy: {0}'.format(e))
+                # We have got a better ernergy value
+                self.state.current_energy = e
+                self.state.current_location = np.copy(x_visit)
+                if e < self.state.ebest:
+                    self.state.ebest = e
+                    self.state.xbest = np.copy(x_visit)
+                    self.state_improved = True
+                    self.not_improved_idx = 0
+            else:
+                # We have not improved but do we accept the new location?
+                r = self._rand_state.random_sample()
+                pqv_temp = (self.acceptance_param - 1.0) * (
+                    e - self.state.current_energy) / self.temperature_step + 1.
+                if pqv_temp < 0.:
+                    pqv = 0.
+                else:
+                    pqv = np.exp(np.log(pqv_temp) / (
+                        1. - self.acceptance_param))
+                if r <= pqv:
+                    # We accept the new location and update state
+                    self.state.current_energy = e
+                    self.state.current_location = np.copy(x_visit)
+                    self.xmin = np.copy(self.state.current_location)
+
+                # No improvement since long time
+                if self.not_improved_idx >= self.not_improved_max_idx:
+                    if j == 0 or self.state.current_energy < self.emin:
+                        self.emin = self.state.current_energy
+                        self.xmin = np.copy(self.state.current_location)
```

Comment:
What is the purpose of this block?  if local_search is disabled emin will only be updated here.  Does that make sense?